### PR TITLE
feat(smartAutocomplete): Go to Definition

### PR DESCRIPTION
**Pull request by hiaux0**
#### 1. From aurelia view to viewModel
```html
${|viewModelVar}
.bind="|otherViewModelVar"
```
-->

```js
class ... {
  @bindable |viewModelVar;

  |otherViewModelVar: any;
}
```

au-ext-goto-view-to-viewModel

#### 2. From view to view, eg. `button` to `button of buttons`
```html
${|button}
```
-->
```html
repeat.for="|button of buttons"
``` 

au-ext-goto-view-to-view

#### 3. Follow custom-element 
```html
<|custom-ele>
<require from="/|custom-ele">
```
-->

```js
|class CustomEle {
}
```

custom-element-name

#### 4. Attribute bindings

```html
<!-- otherHtmlFile.html -->
<require from="/custom-ele">
<custom-ele
  |bind-me-now.bind="..."
></custom-ele>
```
-->

```js
class CustomEle {
  @bindable |bindMeNow;
}
```

attribute-binding

## Motivation
Previously, if I want to find the related class variable in the view, I had to copy paste the value and search in the viewModel.
Similarly `button of buttons`.

## Known issues
- [ ] if different viewModels have same @bindable variable names, definition leads to 'random' viewModel
- [x] 4. limited to `.bind`
  Should include `.call`, `.two-way`, ...

## Todo
- [x] resolve conflict
- [x] cursor always at the end of line
- [x] provied definitions for `my-bindable` in 
```html
<custom-element
  my-bindable.bind="someValue"
>
</custom-element>
```

### CODE REVIEW

**Comment by EisenbergEffect** [comment]
Awesome work @hiaux0 ! I'm sure @eriklieben is very happy to have such a great collaborator 😄

**Comment by hiaux0** [comment]
Thank you @EisenbergEffect! I'm excited to contribute to Aurelia in some way 👨‍💻

File: src/server/ExposeAureliaDefinitions.ts [hunk 1]
```diff
@@ -0,0 +1,256 @@
...
+          )
+        );
+        definitionsInfo[definitionWord] = def;
+      }
+
+      /*
+       * 2.
+       * Here, I assume, that .binding will only be true for valid attr. bindings
+       */
+      if (attr.binding === 'bind') {
```
Comment (hiaux0) [inline src/server/ExposeAureliaDefinitions.ts]: > 4. limited to .bind
Should include .call, .two-way, ...

Should add an array with all possible bindings here

File: src/server/FileParser/Parsers/AureliaHtmlParser.ts [hunk 2]
```diff
@@ -87,15 +87,22 @@ export class AureliaHtmlParser {
...
+
     let bindings = [];
     const aureliaParser = new Parser();
     const interpolationRegex = /\$\{(.*)\}/g;
     var match;
     while (match = interpolationRegex.exec(fileContent)) {
       bindings.push({
         value: match[0],
-        bindingData: aureliaParser.parse(match[1])
+        // bindingData: aureliaParser.parse(match[1]) // some files got parsing errors
```
Comment (hiaux0) [inline src/server/FileParser/Parsers/AureliaHtmlParser.ts]: This line causes some issues in files with nested string interpolations 
(compare #101)

So for the sake of 'goto' feature, I commented that out.
Could uncomment and have few files not work, but I'd rather have goto in 99% of my files

File: src/server/main.ts [hunk 3]
```diff
@@ -99,12 +116,26 @@ connection.onCompletion(async (textDocumentPosition) => {
 });
 
 
-connection.onRequest('aurelia-view-information', (filePath: string) => {
+connection.onRequest('aurelia-view-information', async (filePath: string) => {
+  let fileProcessor = new ProcessFiles();
+  await fileProcessor.processPath();
+  aureliaApplication.components = fileProcessor.components;
```
Comment (hiaux0) [inline src/server/main.ts]: Additionally, to processing files on extension startup.
We need to update `aureliaApplication.components` for new 'gotos'.

This is the fastest way I could provide updates

**Comment by hiaux0** [comment]
@eriklieben There is one todo left, which I am currently not able to resolve

>  if different viewModels have same @bindable variable names, definition leads to 'random' viewModel

The problem with that is, we only get the 'targetWord', eg. `twin-brother`, but here, you can see two components having the same bindable.
It would be ideal to provide the small modal, which vscode gives on mulitple defs, but that is the point I cannot do.
https://github.com/aurelia/vscode-extension/blob/9f0dfe163e79cd2ead19997b82d36ed1e626f050/src/client/main.ts#L66
The return here should accept an array as well, but it does not work.

image

All in all, I feel good about the current state.
Have been already using it locally for weeks now.
(Kind of a workaround for the todo would be, to click the tag name instead of the attribute itself.
You would be presented the correct file, but not the correct variable def)

**Comment by eriklieben** [comment]
awesome! 

Sorry for the late response, my todo list isn't in the state I want it in :-). I will try to merge everything this Sunday and check if we can also get it released to the marketplace because it looks like that isn't happening automatically anymore.

**Comment by bigopon** [comment]
> The problem with that is, we only get the 'targetWord', eg. twin-brother, but here, you can see two components having the same bindable.

Is it unable to read the tagname?

**Comment by hiaux0** [comment]
> Is it unable to read the tagname?

Yes, basically.
But that's just a limitation of the current implementation.
We parse tagnames somehwhere else in the code already, so would be just a matter of adding read-tagname-logic

**Comment by jboysen** [comment]
Is it possible to beta-test this feature? We are looking very much forward to this :)

**Comment by EisenbergEffect** [comment]
@eriklieben What is our path to getting this in and getting it released?

**Comment by eriklieben** [comment]
Thanks for the reminder, sorry this was outside of my sight.

Merged it, will only need to get it released to the marketplace, but that has some issues. The token is expired or there is something with my installation of vsce.

If you want to test it locally, I think you can also do this:
```
git clone repo
npm install
npm i vsce -g
```
then uninstall the extension from vscode, and use the following command to create a package:
```
vsce package
```
(that will build the solution as well if I am correct, otherwise do npm run build as well.)

That will create a vsix file, which you can install locally:
open vscode, go to the extensions tab, and click the ... on the right top, and then "install from VSIX".

(double click on the file won't work)

**Comment by EisenbergEffect** [comment]
@eriklieben Let me know if you need anything from me to help you get this publishing again. I recall I had to refresh the token in the past. Just let me know.

**Pull request merged**